garden_ai/pipelines.py

```diff
@@ -174,14 +179,85 @@ def _collect_requirements(self):
             )
         return
 
-    def register(self):
-        """register this Pipeline's complete Step composition as a globus compute function
-        TODO
+    def __call__(
+        self,
+        *args: Any,
+        endpoint: Union[UUID, str, None] = None,
+        timeout=None,
+        **kwargs: Any,
+    ) -> Any:
+        """Call the pipeline's composed steps on the given input data.
+
+        Attempt to execute remotely if this pipeline has already been registered
+        and a valid ``endpoint`` is specified, else run locally as a regular
+        python function.
+
+        Parameters
+        ----------
+        *args : Any
+            Input data passed through the first step in the pipeline
+        endpoint : Union[UUID, str, None]
+            A valid globus compute endpoint uuid
+        timeout : int
+            time (in seconds) to wait for results. Pass `None` to wait
+            indefinitely (default behavior).
+        **kwargs : Any
+            Additional keyword arguments passed directly to the first step in
+            the pipeline.
+
+        Returns
+        -------
+        Any
+            Results from the pipeline's composed steps called with the given
+            input data.
+
+        Raises
+        ------
+        Exception
+            Any exceptions raised over the course of executing the pipeline
+            function, remotely or otherwise.
+
         """
-        raise NotImplementedError
+        run_locally = not (self.func_uuid and endpoint)
+        polling_interval = 2  # TODO might be a better value for this
 
-    def __call__(self, *args: Any, **kwargs: Any) -> Any:
-        return self._composed_steps(*args, **kwargs)
+        if run_locally:
+            if endpoint:
+                logger.warning(
+                    f"Pipeline '{self.title}' invoked with endpoint="
+                    f" {endpoint}, but has not been registered yet. The "
+                    "pipeline will not execute remotely.",
+                )
+            if self.func_uuid:
+                logger.warning(
+                    f"Pipeline '{self.title}' has been registered for remote "
+                    "execution, but no endpoint was specified. The pipeline will "
+                    "not execute remotely.",
+                )
+            # pass input directly to underlying steps
+            return self._composed_steps(*args, **kwargs)
+        else:
```

Comment:
More and more I'm convinced that "local" pipelines that actually know the contents of their steps and "remote" pipelines that just have the remote id are two quite separate entities. Not something for us to address in this PR, but just pointing at the really disjoint logic here.

Reply:
yeah, I'm definitely on the same page. especially if in the next PR we make it so that you can only run remote pipelines from an authed `client.run(pipeline, ...)` or whatever, might be the perfect time to make that a hard distinction